File: Source/Plugins/Server/com.tle.web.workflow/src/com/tle/web/workflow/notification/TaskNotifications.java
package com.tle.web.workflow.notification;

import java.io.StringWriter;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ListMultimap;
import com.google.common.collect.Lists;
import com.tle.beans.item.Item;
import com.tle.beans.item.ItemId;
import com.tle.beans.item.ItemTaskId;
import com.tle.common.workflow.Workflow;
import com.tle.common.workflow.WorkflowItemStatus;
import com.tle.common.workflow.WorkflowMessage;
import com.tle.common.workflow.WorkflowNodeStatus;
import com.tle.common.workflow.node.WorkflowItem;
import com.tle.common.workflow.node.WorkflowItem.AutoAction;
import com.tle.core.guice.Bind;
import com.tle.core.notification.beans.Notification;
import com.tle.core.services.entity.WorkflowService;
import com.tle.core.services.item.ItemService;
import com.tle.core.services.user.LazyUserLookup;
import com.tle.core.services.user.UserService;
import com.tle.web.i18n.BundleCache;
import com.tle.web.notification.WebNotificationExtension;
import com.tle.web.sections.SectionsController;
import com.tle.web.sections.equella.annotation.PlugKey;
import com.tle.web.sections.equella.freemarker.ExtendedFreemarkerFactory;
import com.tle.web.sections.render.Label;
import com.tle.web.sections.result.util.BundleLabel;
import com.tle.web.sections.result.util.ItemNameLabel;
import com.tle.web.sections.result.util.KeyLabel;
import com.tle.web.sections.result.util.PluralKeyLabel;
import com.tle.web.sections.result.util.UserLabel;
import com.tle.web.workflow.tasks.RootTaskListSection;

@Bind
@Singleton
@SuppressWarnings("nls")
public class TaskNotifications implements WebNotificationExtension
{
	@PlugKey("notereason.")
	private static String KEY_REASON_FILTER;
	@PlugKey("notificationlist.reasons.")
	private static String KEY_REASON_LIST;
	@PlugKey("email.automsg")
	private static String KEY_AUTOMSG;
	@PlugKey("email.autodays")
	private static String KEY_AUTODAYS;
	@PlugKey("email.autoreject")
	private static Label LABEL_AUTOREJECT;
	@PlugKey("email.autoaccept")
	private static Label LABEL_AUTOACCEPT;
	@PlugKey("email.rejecttask")
	private static Label LABEL_CAUSEREJECTED;
	@PlugKey("email.accepttask")
	private static Label LABEL_CAUSEACCEPTED;
	@PlugKey("email.msg.")
	private static String KEY_MESSAGELABEL;

	@Inject
	private BundleCache bundleCache;

	@Inject
	private SectionsController controller;
	@Inject
	private WorkflowService workflowService;
	@Inject
	private ItemService itemService;
	@Inject
	private UserService userService;
	@Inject
	private ExtendedFreemarkerFactory viewFactory;

	@Override
	public boolean isIndexed(String type)
	{
		return !type.equals(Notification.REASON_MODERATE);
	}

	@Override
	public String emailText(ListMultimap<String, Notification> typeMap)
	{
		LazyUserLookup lazyUserLookup = new LazyUserLookup(userService);
		List<TaskNotification> allModerations = createTaskNotifications(typeMap.get(Notification.REASON_MODERATE),
			lazyUserLookup);

		ListMultimap<String, TaskNotification> moderationTypes = ArrayListMultimap.create();
		for( TaskNotification notification : allModerations )
		{
			String group = "other";
			if( notification.getCauseLabel() != null )
			{
				if( notification.isCauseAccepted() )
				{
					group = "accepted";
				}
				else
				{
					group = "rejected";
				}
			}
			moderationTypes.put(group, notification);
		}

		moderationTypes.putAll("overdue",
			createTaskNotifications(typeMap.get(Notification.REASON_OVERDUE), lazyUserLookup));
		EmailNotifications notificationModel = new EmailNotifications(moderationTypes.asMap());

		StringWriter writer = new StringWriter();
		viewFactory.render(viewFactory.createResultWithModel("notification-tasks.ftl", notificationModel), writer);
		return writer.toString();
	}

	private List<TaskNotification> createTaskNotifications(List<Notification> notifications, LazyUserLookup lazyLookup)
	{
		List<ItemTaskId> itemTaskIds = Lists.newArrayList();
		for( Notification notification : notifications )
		{
			itemTaskIds.add(new ItemTaskId(notification.getItemid()));
		}
		LoadingCache<Workflow, Map<String, WorkflowItem>> map = CacheBuilder.newBuilder().build(
			new CacheLoader<Workflow, Map<String, WorkflowItem>>()
			{
				@Override
				public Map<String, WorkflowItem> load(Workflow workflow)
				{
					return workflow.getAllWorkflowItems();
				}
			});

		List<TaskNotification> taskNotifications = Lists.newArrayList();
		Map<ItemId, Item> items = itemService.queryItemsByItemIds(itemTaskIds);
		for( ItemTaskId itemTaskId : itemTaskIds )
		{
			TaskNotification task = new TaskNotification();
			Item item = items.get(ItemId.fromKey(itemTaskId));
			if( item != null )
			{
				Workflow workflow = item.getItemDefinition().getWorkflow();
				Map<String, WorkflowItem> allWorkflowItems = map.getUnchecked(workflow);
				WorkflowItem workflowItem = allWorkflowItems.get(itemTaskId.getTaskId());
				if( workflowItem != null )
				{
					WorkflowItemStatus status = workflowService.getIncompleteStatus(itemTaskId);
					setupCause(lazyLookup, status, task, workflow);
					task.setDueDate(status.getDateDue());
					setupAutoAction(task, workflowItem);
					task.setTaskName(new BundleLabel(workflowItem.getName(), bundleCache));
					task.setItemName(new ItemNameLabel(item, bundleCache));
					task.setLink(RootTaskListSection.createModerateBookmark(controller, itemTaskId));
					taskNotifications.add(task);
				}
			}
		}
		return taskNotifications;
	}

	private void setupAutoAction(TaskNotification task, WorkflowItem workflowItem)
	{
		AutoAction autoAction = workflowItem.getAutoAction();
		if( autoAction != AutoAction.NONE )
		{
			Label actionName = autoAction == AutoAction.REJECT ? LABEL_AUTOREJECT : LABEL_AUTOACCEPT;
			task.setAutoAction(new KeyLabel(KEY_AUTOMSG, actionName, new PluralKeyLabel(KEY_AUTODAYS, workflowItem
				.getActionDays())));
		}
	}

	private void setupCause(LazyUserLookup lazyLookup, WorkflowItemStatus status, TaskNotification task,
		Workflow workflow)
	{
		WorkflowNodeStatus cause = status.getCause();
		if( cause != null )
		{
			if( !workflow.getRoot().getUuid().equals(cause.getNode().getUuid()) )
			{
				boolean accepted = cause.getStatus() == WorkflowNodeStatus.COMPLETE;
				task.setCauseLabel(accepted ? LABEL_CAUSEACCEPTED : LABEL_CAUSEREJECTED);
				task.setCauseTask(new BundleLabel(cause.getNode().getName(), bundleCache));
				task.setCauseAccepted(accepted);
			}
			List<Message> messages = Lists.newArrayList();
			Set<WorkflowMessage> comments = cause.getComments();
			for( WorkflowMessage workflowMessage : comments )
			{
				char type = workflowMessage.getType();
				if( type == WorkflowMessage.TYPE_ACCEPT || type == WorkflowMessage.TYPE_REJECT
					|| type == WorkflowMessage.TYPE_SUBMIT )
				{
					messages.add(new Message(lazyLookup, workflowMessage));
				}
			}
			Collections.sort(messages, new Comparator<Message>()
			{
				@Override
				public int compare(Message m1, Message m2)
				{
					return m1.msg.getDate().compareTo(m2.msg.getDate());
				}
			});
			task.setMessages(messages);
		}
	}

	public static class EmailNotifications
	{
		private final Map<String, Collection<TaskNotification>> groups;

		public EmailNotifications(Map<String, Collection<TaskNotification>> groups)
		{
			this.groups = groups;
		}

		public Map<String, Collection<TaskNotification>> getGroups()
		{
			return groups;
		}
	}

	public static class TaskNotification extends ItemNotification
	{
		private Label taskName;
		private Label causeLabel;
		private Label causeTask;
		private boolean causeAccepted;
		private List<Message> messages = Collections.emptyList();
		private boolean optional;
		private Date dueDate;
		private Label autoAction;

		public Label getTaskName()
		{
			return taskName;
		}

		public void setTaskName(Label taskName)
		{
			this.taskName = taskName;
		}

		public boolean isOptional()
		{
			return optional;
		}

		public void setOptional(boolean optional)
		{
			this.optional = optional;
		}

		public Date getDueDate()
		{
			return dueDate;
		}

		public void setDueDate(Date dueDate)
		{
			this.dueDate = dueDate;
		}

		public Label getAutoAction()
		{
			return autoAction;
		}

		public void setAutoAction(Label autoAction)
		{
			this.autoAction = autoAction;
		}

		public Label getCauseLabel()
		{
			return causeLabel;
		}

		public void setCauseLabel(Label causeLabel)
		{
			this.causeLabel = causeLabel;
		}

		public Label getCauseTask()
		{
			return causeTask;
		}

		public void setCauseTask(Label causeTask)
		{
			this.causeTask = causeTask;
		}

		public List<Message> getMessages()
		{
			return messages;
		}

		public void setMessages(List<Message> messages)
		{
			this.messages = messages;
		}

		public boolean isCauseAccepted()
		{
			return causeAccepted;
		}

		public void setCauseAccepted(boolean causeAccepted)
		{
			this.causeAccepted = causeAccepted;
		}

	}

	public static class Message
	{
		private final WorkflowMessage msg;
		private UserLabel by;

		public Message(LazyUserLookup lazyLookup, WorkflowMessage message)
		{
			this.msg = message;
			this.by = new UserLabel(msg.getUser(), lazyLookup);
		}

		public Label getLabel()
		{
			return new KeyLabel(KEY_MESSAGELABEL + msg.getType());
		}

		public Label getBy()
		{
			return by;
		}

		public String getMessage()
		{
			return msg.getMessage();
		}
	}

	@Override
	public boolean isForceEmail(String type)
	{
		return type.equals(Notification.REASON_OVERDUE);
	}

	@Override
	public Label getReasonLabel(String type)
	{
		return new KeyLabel(KEY_REASON_LIST + type);
	}

	@Override
	public Label getReasonFilterLabel(String type)
	{
		return new KeyLabel(KEY_REASON_FILTER + type);
	}

}

File: Source/Plugins/Workflow/com.tle.web.workflow/scalasrc/com/tle/web/workflow/notification/TaskNotifications.scala
/*
 * Copyright 2017 Apereo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tle.web.workflow.notification

import javax.inject.{Inject, Singleton}

import com.dytech.devlib.PropBagEx
import com.dytech.edge.common.Constants
import com.tle.beans.item.Item
import com.tle.common.NameValue
import com.tle.common.usermanagement.user.valuebean.UserBean
import com.tle.common.workflow.node.WorkflowItem
import com.tle.common.workflow.node.WorkflowItem.AutoAction
import com.tle.common.workflow.{WorkflowMessage, WorkflowNodeStatus}
import com.tle.core.filesystem.WorkflowMessageFile
import com.tle.core.guice.Bind
import com.tle.core.institution.InstitutionService
import com.tle.core.notification.beans.Notification
import com.tle.core.services.FileSystemService
import com.tle.core.services.user.{LazyUserLookup, UserService}
import com.tle.core.workflow.service.WorkflowService
import com.tle.web.sections.render.{AppendedLabel, Label}
import com.tle.web.sections.result.util._
import com.tle.web.workflow.notification.NotificationLangStrings._
import com.tle.web.workflow.servlet.WorkflowMessageServlet

import scala.collection.JavaConverters._

object TaskNotifications {

  import NotificationLangStrings.{l, r}

  val LABEL_CAUSEREJECTED = l("email.rejecttask")
  val LABEL_CAUSEACCEPTED = l("email.accepttask")
  val KEY_MESSAGELABEL = r.key("email.msg.")
  val KEY_FILEMESSAGE = r.key("email.file.")

  val KEY_AUTOMSG = r.key("email.automsg")
  var KEY_AUTODAYS = r.key("email.autodays")
  val LABEL_AUTOREJECT = l("email.autoreject")
  val LABEL_AUTOACCEPT = l("email.autoaccept")
}

import com.tle.web.workflow.notification.TaskNotifications._

@Bind
@Singleton
class TaskNotifications extends FilterableNotification with NotificationLookup with TemplatedNotification {

  type N = TaskNoteModel
  @Inject
  var workflowService: WorkflowService = _
  @Inject
  var userService: UserService = _
  @Inject
  var institutionService: InstitutionService = _
  @Inject
  var fileSystemService: FileSystemService = _

  override def isIndexed(ntype: String) = ntype != Notification.REASON_MODERATE

  class Message(lazyLookup: LazyUserLookup, msg: WorkflowMessage) {
    def getMessage = msg.getMessage

    def getBy = new UserLabel(msg.getUser, lazyLookup)

    def getLabel = new KeyLabel(KEY_MESSAGELABEL + msg.getType)

    def getFileLabel = new KeyLabel(KEY_FILEMESSAGE + msg.getType)

    def date = msg.getDate

    val uuid = msg.getUuid
    val getFiles = fileSystemService.enumerate(new WorkflowMessageFile(uuid), Constants.BLANK, null).map {
      fe => new NameValue(fe.getName, WorkflowMessageServlet.messageUrl(uuid, fe.getName))
    }
    val isHasFiles = getFiles.nonEmpty


  }

  val msgTypes = Set(WorkflowMessage.TYPE_ACCEPT, WorkflowMessage.TYPE_REJECT, WorkflowMessage.TYPE_SUBMIT)

  case class TaskGroup(multiple: Boolean, rejected: Boolean, taskName: String) extends NotificationGroup
  {
    def groupName = s"task.${if (multiple) "m" else "s"}${if (rejected) "r" else "a"}"

    def subjectLabel: Label = new KeyLabel(KEYPFX_EMAIL_SUBJECT+groupName, taskName)

    def templateName: String = "notification-tasks.ftl"

    def headerHello(user: UserBean): Label = NotificationLangStrings.userHeaderLabel(user)

    def headerReason(total: Int): Label = new KeyLabel(NotificationLangStrings.pluralKey(NotificationLangStrings.KEY_HEADER+groupName, total), taskName)
  }

  case class TaskNoteModel(lul: LazyUserLookup)(val note: Notification, val item: Item) extends TaskNotification with OwnerLookup {
    val status = Option(workflowService.getIncompleteStatus(itemTaskId))
    val currentTask: Option[WorkflowItem] = workflowItem(taskId).collect {
      case wi: WorkflowItem => wi
    }
    val getAutoAction = currentTask.flatMap { workflowItem =>
      Option(workflowItem.getAutoAction).collect {
        case AutoAction.ACCEPT => LABEL_AUTOACCEPT
        case AutoAction.REJECT => LABEL_AUTOREJECT
      }.map { l => new KeyLabel(KEY_AUTOMSG, l, new PluralKeyLabel(KEY_AUTODAYS, workflowItem.getActionDays)) }
    }.orNull
    val causeStep = status.flatMap(wis => Option(wis.getCause))
    val causeAccepted = causeStep.filter(_.getNode.getParent != null).map(_.getStatus == WorkflowNodeStatus.COMPLETE)

    def getCauseTask = causeStep.map(wns => new BundleLabel(wns.getNode.getName, bundleCache)).orNull

    def getCauseLabel = causeAccepted.map(if (_) LABEL_CAUSEACCEPTED else LABEL_CAUSEREJECTED).orNull

    val getMessages = causeStep.map(_.getComments.asScala.collect {
      case c if msgTypes(c.getType) => new Message(lul, c)
    }.toSeq.sortBy(_.date)).getOrElse(Seq.empty).asJava

    def isHasCauseInfo = causeStep.isDefined && (causeAccepted.isDefined || !getMessages.isEmpty)

    def getDueDate = status.map(_.getDateDue).orNull

    def isSingleModerator : Boolean =
      currentTask.map(wi => workflowService.getAllModeratorUserIDs(new PropBagEx(item.getItemXml.getXml), wi))
        .exists(_.size() == 1)

    override def group = note.getReason match {
      case Notification.REASON_OVERDUE => StdNotificationGroup("notification-overdue.ftl", Notification.REASON_OVERDUE)
      case _ => TaskGroup(!isSingleModerator, causeAccepted.exists(!_), getTaskName.getText)
    }
  }

  def toFreemarkerModel(notes: Iterable[Notification]) = createDataIgnore(notes, TaskNoteModel(new LazyUserLookup(userService)))
}
